linalg/src/x86_64_fma.rs

```diff
@@ -11,9 +11,78 @@ mod intel;
 pub fn plug(ops: &mut Ops) {
     if is_x86_feature_detected!("fma") {
         ops.mmv_f32 = Box::new(|_, _| mmm::fma_mmm_f32_64x1::mmm());
+
+        ops.mmm_f32 = Box::new(|_, _, n| {
+            if n.is_none() {
+                return mmm::fma_mmm_f32_16x6::mmm();
+            }
+
+            let n = n.unwrap();
+
+            match n {
+                1 => unreachable!("should've been mmv"),
+                2 => return mmm::fma_mmm_f32_40x2::mmm(),
+                3 => return mmm::fma_mmm_f32_32x3::mmm(),
+                4 => return mmm::fma_mmm_f32_24x4::mmm(),
+                5 => return mmm::fma_mmm_f32_16x5::mmm(),
+                6 => return mmm::fma_mmm_f32_16x6::mmm(),
+                8 => return mmm::fma_mmm_f32_8x8::mmm(),
+                _ => {}
+            };
+
+            let scaling_baseline = 60.0;
+            let kernel_normalized_perf = [
+                44.0 / scaling_baseline, // 8x8
+                54.0 / scaling_baseline, // 2x6
+                54.0 / scaling_baseline, // 2x5
+                54.0 / scaling_baseline, // 3x4
+                54.0 / scaling_baseline, // 4x3
+                54.0 / scaling_baseline, // 5x2
+            ];
+
+            fn compute_efficiency(n: usize, kernel_width: usize, scale: f32) -> f32 {
+                let kernel_width = kernel_width as f32;
+                let n = n as f32;
+                let batch_count = (n / kernel_width).ceil();
+                let actual_count = batch_count * kernel_width;
+                let multi_batch_penalty = 1.0 - batch_count / 100.0;
+                n / actual_count * scale * multi_batch_penalty
+            }
+
+            let efficiencies = [
+                compute_efficiency(n, 8, kernel_normalized_perf[0]),
+                compute_efficiency(n, 6, kernel_normalized_perf[1]),
+                compute_efficiency(n, 5, kernel_normalized_perf[2]),
+                compute_efficiency(n, 4, kernel_normalized_perf[3]),
+                compute_efficiency(n, 3, kernel_normalized_perf[4]),
+                compute_efficiency(n, 2, kernel_normalized_perf[5]),
+            ];
+
+            let best_idx = efficiencies.iter().copied().enumerate().fold((0, 0.0), |max, val| {
+                if val.1 > max.1 {
+                    val
+                } else {
+                    max
+                }
+            });
+
+            match best_idx.0 {
+                0 => return mmm::fma_mmm_f32_8x8::mmm(),
+                1 => return mmm::fma_mmm_f32_16x6::mmm(),
+                2 => return mmm::fma_mmm_f32_16x5::mmm(),
+                3 => return mmm::fma_mmm_f32_24x4::mmm(),
+                4 => return mmm::fma_mmm_f32_32x3::mmm(),
+                5 => return mmm::fma_mmm_f32_40x2::mmm(),
+                _ => unreachable!("not a valid index"),
+            }
+        });
```

Comment:
I hope this approach will be good enough on intel. On arm there were too big discrepancies between proc that we went for training a small DNN model to pick teh right model...

Reply:
I think that a DNN will perform better, as my heuristics are very handwavy. But since it performs better in almost all situations; I'm not too bothered right now -- rather do that as a separate PR if we deem it necessary.

Reply:
absolutely.